### ISSUE TITLE
Replace Enumerable.First() with more defensive logic

### ISSUE THREAD

**Issue by JanLenoch** [Kentico/cloud-sample-app-net#15]
In [ContactsController](https://github.com/Kentico/cloud-sample-app-net/blob/master/DancingGoat/Controllers/ContactsController.cs), [HomeController](https://github.com/Kentico/cloud-sample-app-net/blob/master/DancingGoat/Controllers/HomeController.cs), [Home view](https://github.com/Kentico/cloud-sample-app-net/blob/master/DancingGoat/Views/Home/Index.cshtml), [Coffee view](https://github.com/Kentico/cloud-sample-app-net/blob/master/DancingGoat/Views/Product/Coffee.cshtml), [Cafe](https://github.com/Kentico/cloud-sample-app-net/blob/master/DancingGoat/Views/Shared/DisplayTemplates/Cafe.cshtml), and [Product](https://github.com/Kentico/cloud-sample-app-net/blob/master/DancingGoat/Views/Shared/DisplayTemplates/Product.cshtml) display templates, there is the [Enumerable.First()](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.first--1?view=netframework-4.7) being used.

Missing content can break the site.

Replace these calls with a more defensive logic.

(Internal reference: [DEL-971](https://kentico.atlassian.net/browse/DEL-971))

**Comment by Steve-Fenton** [comment]
I notice that the listed areas are using FirstOrDefault now - is more required to close this issue down? If so, I'm happy to contribute.

**Comment by petrsvihlik** [comment]
The issue is still actual but let me rephrase it a little bit. The site should still be functional* if there are no (zero) content items of the following types:
- coffee
- brewer
- accessory
- article
- cafe
- fact about us
- grinder
- hero unit

*functional == the site doesn't throw any exceptions == all exceptions are either handled or prevented by null checks (the content is just not displayed)

**Comment by petrsvihlik** [comment]
if you want to pick it up @Steve-Fenton just leave a note here, please. thanks!

**Comment by Steve-Fenton** [comment]
Hi @petrsvihlik, I'll take a look. Any preference on the test project in terms of unit testing frameworks?

**Comment by petrsvihlik** [comment]
@Steve-Fenton We prefer nunit for .NET Framework and xunit for .NET Core. 

If you need to reference any other library (e.g. for mocking) the rule of thumb is to use libraries that target .NET Standard. So e.g. Moq or FakeItEasy will do.

**Comment by Steve-Fenton** [comment]
Hi @petrsvihlik - I'm hitting a few problems creating mocks for some of the types, as they are sealed and have internal-only constructors, for example `DeliveryItemListingResponse<Coffee>`. Is there any chance of getting an interface over these types... or are there any suggestions for how I could mock these types?

Thanks,

Steve

**Comment by petrsvihlik** [comment]
Hi @Steve-Fenton , yes...it should be possible since yesterday :)

Read more [here](https://forums.kenticocloud.com/discussion/comment/157#Comment_157) and [here](https://github.com/Kentico/delivery-sdk-net/wiki/Faking-responses).

**Comment by Steve-Fenton** [comment]
@petrsvihlik brilliant stuff. I'll update the packages and get working on this. Thanks for the information.

**Comment by Steve-Fenton** [comment]
Hello,

I have been testing this using a web crawler against my local instance, as the amount of changes I would have to make in order to bring the views under test is monumental.

I ran a full crawl, then started un-publishing various content items from my copy of the demo data.

The errors I found from this testing are:

 1. **AboutUsController** Line 21: null Facts -> foreach (var fact in response.Item.Facts)

Fix:

```
        public async Task<ActionResult> Index()
        {
            var response = await client.GetItemAsync<AboutUs>("about_us");

            var viewModel = new AboutUsViewModel
            {
                FactViewModels = MapFactsAboutUs(response)
            };

            return View(viewModel);
        }

        private IList<FactAboutUsViewModel> MapFactsAboutUs(DeliveryItemResponse<AboutUs> response)
        {
            var facts = new List<FactAboutUsViewModel>();

            if (response.Item == null)
            {
                return facts;
            }

            int i = 0;

            foreach (var fact in response.Item.Facts)
            {
                var factViewModel = new FactAboutUsViewModel
                {
                    Fact = (FactAboutUs)fact
                };

                if (i++ % 2 == 0)
                {
                    factViewModel.Odd = true;
                }

                facts.Add(factViewModel);
            }

            return facts;
        }
```

 2. **Contacts/Index.cshtml** Line 10: null Roastery -> &lt;li>@Html.DisplayFor(vm => vm.Roastery.Phone)&lt;/li>

Fixed by wrapping with:

`@if (Model.Roastery != null)
{`

 3. Shared/DisplayTemplates/PhoneNumber.cshtml: null string

Before

`string strippedDashes = Model.Replace("-", string.Empty);`

After

`string strippedDashes = string.IsNullOrWhiteSpace(Model) ? string.Empty : Model.Replace("-", string.Empty);`

I am currently dealing with a combination of VS2015/GitHub two-factor auth problems that is preventing this local change going up to my branch, and having to look at a live issue in my day job, so I don't know when the commit is coming for this.

**Comment by petrsvihlik** [comment]
@Steve-Fenton last thing before we close this issue. Could you please verify the following three scenarios?

- When home page is missing Hero image, it ends up with an unhandled exception
- When home page is missing content variants for personalization, it ends up with unhandled exception
- ContactsControler - If there are no Cafes, throws exception on .First()

They were reported some time ago so I'm not sure if they're still valid.

**Comment by Steve-Fenton** [comment]
* When home page is missing Hero image, it ends up with an unhandled exception

I have tested using "unpublish" on "Home page hero unit" and there are no errors.

![no-hero](https://user-images.githubusercontent.com/1036595/29611742-a65a971e-87f6-11e7-9a8d-601f52037a42.png)

 * When home page is missing content variants for personalization, it ends up with unhandled except

I'm not sure which content to remove on this one. Can I have a hint?

 * ContactsControler - If there are no Cafes, throws exception on .First()

Again, I unpublished all cafes. There is no error, although the page looks very blank.

![no-cafe](https://user-images.githubusercontent.com/1036595/29611743-ab45c988-87f6-11e7-8337-657e12b784c0.png)
